### PR TITLE
Add compatibility info option

### LINKED ISSUE #521: Add option to get reign for ytserver-all
It would be useful for https://github.com/ytsaurus/yt-k8s-operator/issues/200 to compare old and new image of master reign.

Something like `/usr/bin/ytserver-all --reign` would do.

### PR DESCRIPTION
**Pull request by abogutskiy**
Resolves #521

### CODE REVIEW

**Comment by abogutskiy** [comment]
```
ee8jsmrbc5bm206k836c@bh2b070bfacvcc95n4tt:~/ytsaurus/yt/yt/server/all$ ./ytserver-all ytserver-master --compatibility-info
{
    "current_reign" = 2618;
}
```

File: yt/yt/server/master/cell_master/program.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
 #include "snapshot_exporter.h"
 
 #include <yt/yt/server/lib/hydra/dry_run/utils.h>
+#include <yt/yt/server/master/cell_master/serialize.h>
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:6]: This file is located in the same directory, so just `#include "serialize.h"`

File: yt/yt/server/master/cell_master/program.h [hunk 2]
```diff
@@ -17,6 +18,7 @@
 #include <yt/yt/core/bus/tcp/dispatcher.h>
 
 #include <yt/yt/core/misc/shutdown.h>
+#include <yt/yt/core/yson/writer.h>
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:21]: We separate includes from differents directories with an empty line.

File: yt/yt/server/master/cell_master/program.h [hunk 3]
```diff
@@ -87,13 +93,15 @@ class TCellMasterProgram
...
         auto replayChangelogs = parseResult.Has("replay-changelogs");
         auto buildSnapshot = parseResult.Has("build-snapshot");
 
-        if (dumpSnapshot + validateSnapshot + exportSnapshot > 1) {
-            THROW_ERROR_EXCEPTION("Options 'dump-snapshot', 'validate-snapshot' and 'export-snapshot' are mutually exclusive");
+        if (dumpSnapshot + validateSnapshot + exportSnapshot + printCompatibilityInfo  > 1) {
+            THROW_ERROR_EXCEPTION("Options 'dump-snapshot', 'validate-snapshot', 'export-snapshot' and 'compatibility-info' are mutually exclusive");
         }
 
+
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:104]: We rarely separate blocks with two empty lines, once is enough here.

File: yt/yt/server/master/cell_master/program.h [hunk 4]
```diff
@@ -102,6 +110,11 @@ class TCellMasterProgram
             THROW_ERROR_EXCEPTION("Option 'build-snapshot' can only be used with 'validate-snapshot' or 'replay-changelog'");
         }
 
+
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:113]: Also here.

File: yt/yt/server/master/cell_master/program.h [hunk 5]
```diff
@@ -201,13 +214,27 @@ class TCellMasterProgram
         _exit(0);
     }
 
+private:
+
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:218]: There is no need in blank line between `private:` and first method

File: yt/yt/server/master/cell_master/program.h [hunk 6]
```diff
@@ -201,13 +214,27 @@ class TCellMasterProgram
...
+    void PrintCompatibilityInfoAndExit() const {
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:219]: For functions and classes, `{` are put onto the next line (and for ifs and fors no):

```
void MaybeExit(bool exit)
{
    if (exit) {
        _exit(0);
    }
}
```

File: yt/yt/server/master/cell_master/program.h [hunk 7]
```diff
@@ -201,13 +213,27 @@ class TCellMasterProgram
         _exit(0);
     }
 
+private:
+    void PrintCompatibilityInfoAndExit() const
+    {
+        NYT::NYson::TYsonWriter writer(&Cout, NYson::EYsonFormat::Pretty);
```
Comment (gritukan) [inline yt/yt/server/master/cell_master/program.h:219]: NYT:: here and below is not required

**Review by gritukan — APPROVED** [review]

File: yt/yt/server/master/cell_master/program.h [hunk 8]
```diff
@@ -201,13 +213,27 @@ class TCellMasterProgram
         _exit(0);
     }
 
+private:
+    void PrintCompatibilityInfoAndExit() const
+    {
+        NYson::TYsonWriter writer(&Cout, NYson::EYsonFormat::Pretty);
+        auto node = NYTree::BuildYsonNodeFluently()
```
Comment (maxim-babenko) [inline yt/yt/server/master/cell_master/program.h:220]: Let's use BuildYsonStringFluently

**Comment by maxim-babenko** [comment]
Please capitalize commit message

File: yt/yt/server/master/cell_master/program.h [hunk 9]
```diff
@@ -201,13 +213,27 @@ class TCellMasterProgram
         _exit(0);
     }
 
+private:
+    void PrintCompatibilityInfoAndExit() const
```
Comment (maxim-babenko) [inline yt/yt/server/master/cell_master/program.h:217]: const makes little sense here, let's drop it

File: yt/yt/server/master/cell_master/program.h [hunk 10]
```diff
@@ -102,6 +110,10 @@ class TCellMasterProgram
             THROW_ERROR_EXCEPTION("Option 'build-snapshot' can only be used with 'validate-snapshot' or 'replay-changelog'");
         }
 
+        if (PrintCompatibilityInfo_) {
```
Comment (maxim-babenko) [inline yt/yt/server/master/cell_master/program.h:113]: Below we have a different "HandleSomething" pattern. Can we follow it here as well?

File: yt/yt/server/master/cell_master/program.h [hunk 11]
```diff
@@ -201,13 +213,27 @@ class TCellMasterProgram
         _exit(0);
     }
 
+private:
+    void PrintCompatibilityInfoAndExit() const
+    {
+        NYson::TYsonWriter writer(&Cout, NYson::EYsonFormat::Pretty);
+        auto node = NYTree::BuildYsonNodeFluently()
+          .BeginMap()
+              .Item("current_reign").Value(static_cast<int>(NCellMaster::GetCurrentReign()))
```
Comment (maxim-babenko) [inline yt/yt/server/master/cell_master/program.h:222]: ToUnderlying looks better

File: yt/yt/server/master/cell_master/program.h (hunk 10 above)
Comment (abogutskiy) [reply yt/yt/server/master/cell_master/program.h:113]: Would it be enough to change method to `bool HandleCompatibilityInfo()`? Or do you want some extendable generic class for compatibility info which is enriched with reign for cell-master?

File: yt/yt/server/master/cell_master/program.h (hunk 11 above)
Comment (abogutskiy) [reply yt/yt/server/master/cell_master/program.h]: actually it was TReign already, removed unnecessary cast

**Comment by maxim-babenko** [comment]
Shipped!

File: yt/yt/server/master/cell_master/program.h [hunk 12]
```diff
@@ -201,13 +213,32 @@ class TCellMasterProgram
...
+    {
+        if (!PrintCompatibilityInfo_) {
+            return false;
+        }
+
+        NYson::TYsonWriter writer(&Cout, NYson::EYsonFormat::Pretty);
+        auto info = NYTree::BuildYsonStringFluently()
+          .BeginMap()
+              .Item("current_reign").Value(NCellMaster::GetCurrentReign())
+          .EndMap();
```
Comment (Kontakter) [inline yt/yt/server/master/cell_master/program.h:227]: There is wrong identation here. Let fix it with separate PR.

**Pull request closed without merging**